### PR TITLE
feat: Snapshot checksum calculation and verification

### PR DESCRIPTION
**Pull request by derekbit**
https://github.com/longhorn/longhorn/issues/4210
https://github.com/longhorn/longhorn/issues/3198

Signed-off-by: Derek Su [derek.su@suse.com](mailto:derek.su@suse.com)

### CODE REVIEW

File: pkg/client/proxy_snapshot.go [hunk 1]
```diff
@@ -5,12 +5,10 @@ import (
 
 	etypes "github.com/longhorn/longhorn-engine/pkg/types"
 	eutil "github.com/longhorn/longhorn-engine/pkg/util"
+	ediskutil "github.com/longhorn/longhorn-engine/pkg/util/disk"
```
Comment (shuo-wu) [inline pkg/client/proxy_snapshot.go:8]: Move const `VolumeHeadName` to another file like `"github.com/longhorn/longhorn-engine/pkg/types/types"` then we can use alias `enginetypes`

File: pkg/imrpc/proxy.proto [hunk 2]
```diff
@@ -214,6 +216,13 @@ message EngineReplicaRemoveRequest {
 	string replica_address = 2;
 }
 
+message EngineReplicaUpdateRequest {
```
Comment (shuo-wu) [inline pkg/imrpc/proxy.proto:219]: 1. This is more like `ReplicaModeUpdate`
2. The field `mode` is a controller in-memory field. Should replicas understand it?

File: pkg/client/proxy_snapshot.go (hunk 1 above)
Comment (derekbit) [reply pkg/client/proxy_snapshot.go:8]: Sound good. Thank you.

File: pkg/imrpc/proxy.proto (hunk 2 above)
Comment (derekbit) [reply pkg/imrpc/proxy.proto]: Updated. Thank you.

**Comment by derekbit** [comment]
Please help review. Thank you.
cc @shuo-wu @c3y1huang @innobead

**Review by shuo-wu — DISMISSED** [review]
LGTM if there is no doubt about the engine part

**Comment by derekbit** [comment]
Will update dependencies if longhorn-engine PR is merged.

**Comment by derekbit** [comment]
This is ready for review.
cc @innobead

**Review by innobead — APPROVED** [review]

**Pull request merged**